common/finalizer/finalizer.go

```diff
@@ -0,0 +1,170 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package finalizer
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/internal/goro"
+)
+
+var (
+	FinalizerAlreadyDoneErr = errors.New("finalizer already finalized")
+	FinalizerUnknownIdErr   = errors.New("finalizer callback not found")
+	FinalizerDuplicateIdErr = errors.New("finalizer callback already registered")
+)
+
+type Finalizer struct {
+	logger    log.Logger
+	mu        sync.Mutex  // lock protects access to callbacks map
+	finalized atomic.Bool // indicating if finalizer has been executed
+	callbacks map[string]func(context.Context) error
+}
+
+func NewFinalizer(
+	logger log.Logger,
+) *Finalizer {
+	return &Finalizer{
+		logger:    logger,
+		callbacks: make(map[string]func(context.Context) error),
+	}
+}
+
+// Register adds a callback to the finalizer.
+// Returns an error if the ID is already registered, or when the finalizer already ran.
+func (f *Finalizer) Register(
+	id string,
+	callback func(context.Context) error,
+) error {
+	if f.finalized.Load() {
+		return FinalizerAlreadyDoneErr
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if _, ok := f.callbacks[id]; ok {
+		return FinalizerDuplicateIdErr
+	}
+	f.callbacks[id] = callback
+	return nil
+}
+
+// Deregister removes a callback from the finalizer.
+// Returns an error if the ID is not found, or when the finalizer already ran.
+func (f *Finalizer) Deregister(
+	id string,
+) (err error) {
+	if f.finalized.Load() {
+		return FinalizerAlreadyDoneErr
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if _, ok := f.callbacks[id]; !ok {
+		return FinalizerUnknownIdErr
+	}
+	delete(f.callbacks, id)
+	return nil
+}
+
+// Run executes all registered callback functions within the given timeout (zero timeout skips execution).
+// It can only be invoked once; calling it again has no effect.
+// Returns the number of completed callbacks.
+func (f *Finalizer) Run(
+	pool *goro.AdaptivePool,
+	timeout time.Duration,
+) int32 {
+	if !f.finalized.CompareAndSwap(false, true) {
+		f.logger.Warn("finalizer skipped: called more than once")
+		return 0
+	}
+
+	if timeout == 0 {
+		f.logger.Warn("finalizer skipped: zero timeout")
+		return 0
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	totalCount := int32(len(f.callbacks))
+	f.logger.Info("finalizer starting",
+		tag.NewInt32("items", totalCount),
+		tag.NewDurationTag("timeout", timeout))
+
+	completionChannel := make(chan struct{})
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	go func() {
+		for _, callback := range f.callbacks {
+			// NOTE: Once the pool is stopped due to a timeout, any remaining callbacks will not be invoked anymore.
+			pool.Do(func() {
+				defer func() { completionChannel <- struct{}{} }()
+				_ = callback(ctx)
+			})
+		}
+	}()
```

Comment:
this goroutine will continue to make callback even after the timeout?

Reply:
So the way the `pool.Do` works is that it will block once there are not enough workers to handle the work. But also, once you call `Stop`, all further `Do` calls become a noop. It will still finish in-flight work, but after that, it will not execute any of the remaining callbacks.

From its documentation:
```
// Do calls f() on a worker goroutine. If the call can't be started within targetDelay, it adds
// another worker. If Stop is called concurrently, Do may or may not call f. If Stop has been
// called already, Do does nothing. 
```